python/extensions/python.bzl

```diff
@@ -17,35 +17,117 @@
 load("@rules_python//python:repositories.bzl", "python_register_toolchains")
 load("@rules_python//python/extensions/private:interpreter_hub.bzl", "hub_repo")
 
+# Printing a warning msg not debugging, so we have to disable
+# the buildifier check.
+# buildifier: disable=print
+def _print_warn(msg):
+    print("WARNING:", msg)
+
+def _python_register_toolchains(toolchain_attr, version_constraint):
+    python_register_toolchains(
+        name = toolchain_attr.name,
+        python_version = toolchain_attr.python_version,
+        register_coverage_tool = toolchain_attr.configure_coverage_tool,
+        ignore_root_user_error = toolchain_attr.ignore_root_user_error,
+        set_python_version_constraint = version_constraint,
+    )
+
 def _python_impl(module_ctx):
-    toolchains = []
+    # We collect all of the toolchain names to create
+    # the INTERPRETER_LABELS map.  This is used
+    # by interpreter_extensions.bzl via the hub_repo call below.
+    toolchain_names = []
+
+    # Used to store the default toolchain name so we can pass it to the hub
+    default_toolchain_name = None
+
+    # Used to store toolchains that are in sub modules.
+    sub_toolchains_map = {}
+
     for mod in module_ctx.modules:
         for toolchain_attr in mod.tags.toolchain:
-            python_register_toolchains(
-                name = toolchain_attr.name,
-                python_version = toolchain_attr.python_version,
-                bzlmod = True,
-                # Toolchain registration in bzlmod is done in MODULE file
-                register_toolchains = False,
-                register_coverage_tool = toolchain_attr.configure_coverage_tool,
-                ignore_root_user_error = toolchain_attr.ignore_root_user_error,
-            )
-
-            # We collect all of the toolchain names to create
-            # the INTERPRETER_LABELS map.  This is used
-            # by interpreter_extensions.bzl
-            toolchains.append(toolchain_attr.name)
+            # If we are in the root module we always register the toolchain.
+            # We wait to register the default toolchain till the end.
+            if mod.is_root:
+                toolchain_names.append(toolchain_attr.name)
+
+                # If we have the default version or we only have one toolchain
+                # in the root module we set the toolchain as the default toolchain.
+                if toolchain_attr.is_default or len(mod.tags.toolchain) == 1:
+                    # We have already found one default toolchain, and we can
+                    # only have one.
+                    if default_toolchain_name != None:
+                        fail("""We found more than one toolchain that is marked 
```

Comment:
Consider printing to the user the toolchains that are set default to make fixing of the error quicker.